### ISSUE TITLE
FunctionNaming should not report when the function name is identical to the type of the result.

### ISSUE THREAD

**Issue by jcornaz** [arturbosch/detekt#1639]
Currently, [FunctionNaming](https://arturbosch.github.io/detekt/naming.html#functionnaming) reports names which are valid according to the [Kotlin Coding Conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)

## Expected Behavior of the rule
I do not expect any issue report for the following code block, since it comes straight from the official conventions:

```kotlin
abstract class Foo

class FooImpl : Foo()

fun Foo(): Foo { return FooImpl() }
```

Source: https://kotlinlang.org/docs/reference/coding-conventions.html#function-names

But, currently, it reports a `FunctionNaming` because `fun Foo(): Foo` starts with a capital.

## Context

As stated in the conventions (see link above): 
>factory functions used to create instances of classes can have the same name as the class being created:

And I personally find that very useful. If we read `val x = Foo()`, we immediately understand the important point: "It creates an instance of type `Foo`". Whether `Foo()` it is an actual constructor or a factory function, is an implementation detail, and the reader does not have to care.

**Comment by arturbosch** [comment]
Hm, when I copy paste that code into my IDE, `fun Foo` gets flagged for not starting with a lower case letter and that there is a naming conflict for `abstract class Foo` and `fun Foo`.
So for this to work I expect `Foo` coming from another file and the `fun Foo` has `Foo` as an typealias or qualified name?

**Comment by jcornaz** [comment]
>Hm, when I copy paste that code into my IDE, fun Foo gets flagged for not starting with a lower case letter and that there is a naming conflict for abstract class Foo and fun Foo

True, the example from the kotlin conventions does not actually compile. Here is a compilable example:

```
interface Foo

private class FooImpl : Foo

fun Foo(): Foo = FooImpl()
```

**Comment by arturbosch** [comment]
Thanks for the example.
When I copy this lines even IntelliJ complains about the naming, hm.
We still could implement an exception when `methodName == ReturnType`.

**Comment by jcornaz** [comment]
>When I copy this lines even IntelliJ complains about the naming, hm.

Yes IntelliJ also (incorrectly) reports a style issue.

>We still could implement an exception when methodName == ReturnType.

Yes, it'd be nice.
